pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarMetadataEventSynchronizer.java

```diff
@@ -181,19 +238,73 @@ private void startConsumer() {
     }
 
     public boolean isStarted() {
-        return started;
+        return this.state == State.Started;
+    }
+
+    public boolean isProducerStarted() {
+        return this.state.ordinal() > State.Starting_Producer.ordinal()
+                && this.state.ordinal() < State.Closing.ordinal();
+    }
+
+    public boolean isClosingOrClosed() {
+        return this.state == State.Closing || this.state == State.Closed;
     }
 
     @Override
-    public void close() {
-        started = false;
-        if (producer != null) {
-            producer.closeAsync();
-            producer = null;
+    public synchronized CompletableFuture<Void> closeAsync() {
+        while (true) {
```

Comment:
I wonder if it makes sense to avoid `while (true)` loops. One possible improvement:
```suggestion
        while (!Thread.currentThread().isInterrupted()) {
```

Is there a chance to a tight spinning loop which would cause 100% CPU consumption here? Is the loop needed at all?

If it's needed, there would need to be a short delay (100ms?) at the end of each loop.

Reply:
> Is there a chance to a tight spinning loop which would cause 100% CPU consumption here? Is the loop needed at all?

It will never occur, because it checked all the states, the loop at most will be executed 2 times. I added a limitation to limit the max try times to less than `100`